src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerJsonProcessor.java

```diff
@@ -64,25 +72,50 @@
 
   private static final String URL_KEY = "url";
 
+  private static final String NAME_KEY = "name";
+
+  private static final String VERSION_KEY = "version";
+
+  private static final String TIME_KEY = "time";
+
+  private static final String UID_KEY = "uid";
+
   private static final String ZIP_TYPE = "zip";
 
   private static final ObjectMapper mapper = new ObjectMapper();
 
+  private static final DateTimeFormatter timeFormatter = DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ssZZ");
   /**
    * Generates a packages.json file (inclusive of all projects) based on the list.json provided as a payload. Expected
    * usage is to "go remote" on the current repository to fetch a list.json copy, then pass it to this method to build
    * the packages.json for the client to use.
    */
-  public Content generatePackagesJson(final Repository repository, final Payload payload) throws IOException {
+  public Content generatePackagesFromList(final Repository repository, final Payload payload) throws IOException {
     // TODO: Parse using JSON tokens rather than loading all this into memory, it "should" work but I'd be careful.
     Map<String, Object> listJson = parseJson(payload);
-    Collection<String> packageNames = (Collection<String>) listJson.get(PACKAGE_NAMES_KEY);
-    Map<String, Map<String, Object>> packages = packageNames.stream()
-        .collect(Collectors.toMap((each) -> each, (each) -> Collections.singletonMap(SHA256_KEY, null)));
+    return buildPackagesJson(repository, (Collection<String>) listJson.get(PACKAGE_NAMES_KEY));
+  }
+
+  /**
+   * Generates a packages.json file (inclusive of all projects) based on the components provided. Expected usage is
+   * for a hosted repository to be queried for its components, which are then provided to this method to build the
+   * packages.json for the client to use.
+   */
+  public Content generatePackagesFromComponents(final Repository repository, final Iterable<Component> components)
+      throws IOException
+  {
+    return buildPackagesJson(repository, StreamSupport.stream(components.spliterator(), false)
+        .map(component -> component.group() + "/" + component.name()).collect(Collectors.toList()));
+  }
 
+  /**
+   * Builds a packages.json file as a {@code Content} instance containing the actual JSON for the given providers.
+   */
+  private Content buildPackagesJson(final Repository repository, final Collection<String> names) throws IOException {
     Map<String, Object> packagesJson = new LinkedHashMap<>();
     packagesJson.put(PROVIDERS_URL_KEY, repository.getUrl() + "/p/%package%.json");
```

Comment:
Trivial: Maybe make "/p/%package%.json" a constant

Reply:
> Trivial: Maybe make "/p/%package%.json" a constant

Not a bad idea. https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/12/commits/6ea08350768d6d66870aa70d8aece4158d6b4103